File: app/src/main/java/com/codylimber/fieldphenology/data/generator/DatasetGenerator.kt
package com.codylimber.fieldphenology.data.generator

import android.content.Context
import com.codylimber.fieldphenology.data.api.INatApiClient
import com.codylimber.fieldphenology.data.model.*
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import java.io.File
import java.time.Instant

enum class GenerationPhase {
    FETCHING_SPECIES, FETCHING_HISTOGRAMS, FETCHING_DETAILS,
    DOWNLOADING_PHOTOS, SAVING
}

data class GenerationProgress(
    val phase: GenerationPhase,
    val current: Int,
    val total: Int,
    val message: String
)

class DatasetGenerator(
    private val apiClient: INatApiClient,
    private val context: Context
) {
    suspend fun generate(
        placeIds: List<Int>,
        placeName: String,
        taxonIds: List<Int?>,
        taxonName: String,
        groupName: String,
        minObs: Int = 1,
        qualityGrade: String = "research",
        maxPhotos: Int = 3,
        onProgress: (GenerationProgress) -> Unit
    ): String {
        val slug = "${groupName}-${placeName}".lowercase().replace(Regex("[^a-z0-9]+"), "-").trim('-')
        val outputDir = File(context.filesDir, "datasets/$slug")
        val photosDir = File(outputDir, "photos")
        photosDir.mkdirs()

        // Step 1: Fetch species list across all taxon+place combos
        onProgress(GenerationProgress(GenerationPhase.FETCHING_SPECIES, 0, 1, "Fetching species list..."))
        // Map taxonId -> JsonObject, merge by taxon_id summing counts
        val speciesByTaxonId = mutableMapOf<Int, JsonObject>()
        val obsCounts = mutableMapOf<Int, Int>()

        for (taxonId in taxonIds) {
            for (placeId in placeIds) {
                var page = 1
                while (true) {
                    val (total, results) = apiClient.getSpeciesCounts(taxonId, placeId, qualityGrade = qualityGrade, page = page)
                    for (r in results) {
                        val tid = r["taxon"]?.jsonObject?.get("id")?.jsonPrimitive?.intOrNull ?: continue
                        val count = r["count"]?.jsonPrimitive?.intOrNull ?: 0
                        if (tid !in speciesByTaxonId) {
                            speciesByTaxonId[tid] = r
                            obsCounts[tid] = count
                        } else {
                            obsCounts[tid] = (obsCounts[tid] ?: 0) + count
                        }
                    }
                    onProgress(GenerationProgress(GenerationPhase.FETCHING_SPECIES,
                        speciesByTaxonId.size, speciesByTaxonId.size,
                        "Found ${speciesByTaxonId.size} species so far..."))
                    if (results.size < 200 || results.isEmpty()) break
                    page++
                }
            }
        }

        // Filter by min obs (using merged counts)
        val filteredIds = obsCounts.filter { it.value >= minObs }.keys
        val filteredRaw = filteredIds.mapNotNull { tid ->
            speciesByTaxonId[tid]?.let { obj ->
                // Rebuild with merged count
                JsonObject(obj.toMutableMap().apply {
                    put("count", JsonPrimitive(obsCounts[tid] ?: 0))
                })
            }
        }

        val speciesList = DataProcessor.buildSpeciesList(filteredRaw)
        if (speciesList.isEmpty()) throw RuntimeException("No species found matching criteria.")

        val nSpecies = speciesList.size
        onProgress(GenerationProgress(GenerationPhase.FETCHING_SPECIES, nSpecies, nSpecies,
            "$nSpecies species found"))

        // Step 2: Fetch histograms — sum across all places
        val histograms = mutableMapOf<String, Map<Int, Int>>()
        for ((i, sp) in speciesList.withIndex()) {
            onProgress(GenerationProgress(GenerationPhase.FETCHING_HISTOGRAMS, i + 1, nSpecies,
                "Histogram: ${sp.commonName.ifEmpty { sp.scientificName }} (${i + 1}/$nSpecies)"))

            val combined = mutableMapOf<Int, Int>()
            for (placeId in placeIds) {
                val h = apiClient.getHistogram(sp.taxonId, placeId, qualityGrade = qualityGrade)
                for ((week, count) in h) {
                    combined[week] = (combined[week] ?: 0) + count
                }
            }
            histograms[sp.scientificName] = combined
        }

        val weekly = DataProcessor.buildWeeklyMatrix(histograms)

        // Update peak/first/last week
        for (sp in speciesList) {
            val h = histograms[sp.scientificName] ?: continue
            val active = h.filter { it.value > 0 }
            if (active.isNotEmpty()) {
                sp.peakWeek = active.maxByOrNull { it.value }!!.key
                sp.firstWeek = active.keys.min()
                sp.lastWeek = active.keys.max()
            }
            sp.periodCount = DataProcessor.detectFlightPeriods(weekly, sp.scientificName)
        }

        // Step 3: Fetch taxa details
        onProgress(GenerationProgress(GenerationPhase.FETCHING_DETAILS, 0, nSpecies, "Fetching species details..."))
        val taxaDetails = mutableMapOf<Int, JsonObject>()
        val allTaxonIds = speciesList.map { it.taxonId }
        for (i in allTaxonIds.indices step 30) {
            val batch = allTaxonIds.subList(i, minOf(i + 30, allTaxonIds.size))
            val results = apiClient.getTaxaDetails(batch)
            for (t in results) {
                val id = t["id"]?.jsonPrimitive?.intOrNull ?: continue
                taxaDetails[id] = t
            }
            onProgress(GenerationProgress(GenerationPhase.FETCHING_DETAILS,
                minOf(i + 30, allTaxonIds.size), nSpecies,
                "Details: ${minOf(i + 30, allTaxonIds.size)}/$nSpecies"))
        }

        // Step 4: Download photos
        val photoMap = mutableMapOf<Int, List<SpeciesPhoto>>()
        for ((idx, sp) in speciesList.withIndex()) {
            onProgress(GenerationProgress(GenerationPhase.DOWNLOADING_PHOTOS, idx + 1, nSpecies,
                "Photos: ${sp.commonName.ifEmpty { sp.scientificName }} (${idx + 1}/$nSpecies)"))

            val taxonInfo = taxaDetails[sp.taxonId]
            val photos = extractPhotos(taxonInfo)
            val spPhotos = mutableListOf<SpeciesPhoto>()

            // Filter to Creative Commons licensed photos only
            val ccPhotos = photos.filter { photo ->
                val license = photo["license_code"]?.jsonPrimitive?.contentOrNull
                license != null && license.startsWith("cc")
            }

            // Fallback to observation photos if taxon photos are insufficient
            val allCcPhotos = if (ccPhotos.size < maxPhotos) {
                val obsPhotos = try {
                    apiClient.getObservationPhotos(sp.taxonId, placeIds, qualityGrade, maxPhotos - ccPhotos.size)
                } catch (_: Exception) { emptyList() }
                ccPhotos + obsPhotos
            } else {
                ccPhotos
            }

            for ((pi, photo) in allCcPhotos.take(maxPhotos).withIndex()) {
                var url = photo["medium_url"]?.jsonPrimitive?.contentOrNull
                    ?: photo["url"]?.jsonPrimitive?.contentOrNull
                    ?: continue
                url = url.replace("/square.", "/medium.")

                val filename = "${sp.taxonId}_$pi.jpg"
                val bytes = apiClient.downloadPhoto(url) ?: continue
                File(photosDir, filename).writeBytes(bytes)

                spPhotos.add(SpeciesPhoto(
                    file = filename,
                    attribution = photo["attribution"]?.jsonPrimitive?.contentOrNull,
                    license = photo["license_code"]?.jsonPrimitive?.contentOrNull
                ))
            }
            photoMap[sp.taxonId] = spPhotos
        }

        // Step 5: Build and save dataset
        onProgress(GenerationProgress(GenerationPhase.SAVING, 0, 1, "Saving dataset..."))

        val weeklyBySpecies = mutableMapOf<String, MutableList<WeeklyEntry>>()
        for (row in weekly) {
            weeklyBySpecies.getOrPut(row.species) { mutableListOf() }
                .add(WeeklyEntry(row.week, row.n, row.relAbundance))
        }

        val speciesEntries = speciesList.map { sp ->
            val taxonInfo = taxaDetails[sp.taxonId]
            val family = extractAncestorByRank(taxonInfo, "family")
            val familySci = extractAncestorScientificByRank(taxonInfo, "family")
            val order = extractAncestorByRank(taxonInfo, "order")
            val orderSci = extractAncestorScientificByRank(taxonInfo, "order")

            val conservation = safeJsonObject(taxonInfo?.get("conservation_status"))
            var description = taxonInfo?.get("wikipedia_summary")?.jsonPrimitive?.contentOrNull ?: ""
            description = description.replace(Regex("<[^>]+>"), "").trim()

            val spWeekly = weeklyBySpecies[sp.scientificName] ?: mutableListOf()
            val existingWeeks = spWeekly.map { it.week }.toSet()
            for (w in 1..53) {
                if (w !in existingWeeks) spWeekly.add(WeeklyEntry(w, 0, 0f))
            }
            spWeekly.sortBy { it.week }

            Species(
                taxonId = sp.taxonId,
                scientificName = sp.scientificName,
                commonName = sp.commonName,
                totalObs = sp.totalObs,
                rarity = sp.rarity,
                peakWeek = sp.peakWeek,
                firstWeek = sp.firstWeek,
                lastWeek = sp.lastWeek,
                periodCount = sp.periodCount,
                description = description,
                conservationStatus = conservation?.get("status")?.jsonPrimitive?.contentOrNull,
                conservationStatusName = conservation?.get("status_name")?.jsonPrimitive?.contentOrNull,
                family = family,
                familyScientific = familySci,
                order = order,
                orderScientific = orderSci,
                photos = photoMap[sp.taxonId] ?: emptyList(),
                weekly = spWeekly
            )
        }

        val dataset = Dataset(
            metadata = DatasetMetadata(
                placeName = placeName,
                placeId = placeIds.first(),
                placeIds = placeIds,
                group = groupName,
                taxonName = taxonName,
                taxonIds = taxonIds.filterNotNull(),
                totalObs = speciesList.sumOf { it.totalObs },
                speciesCount = speciesList.size,
                generatedAt = Instant.now().toString(),
                minObs = minObs,
                qualityGrade = qualityGrade,
                maxPhotos = maxPhotos
            ),
            species = speciesEntries
        )

        val jsonStr = Json { encodeDefaults = true }.encodeToString(dataset)
        File(outputDir, "dataset.json").writeText(jsonStr)

        onProgress(GenerationProgress(GenerationPhase.SAVING, 1, 1,
            "Complete! ${speciesEntries.size} species"))

        return outputDir.absolutePath
    }

    private fun safeJsonObject(element: JsonElement?): JsonObject? {
        if (element == null || element is JsonNull) return null
        return try { element.jsonObject } catch (_: Exception) { null }
    }

    private fun extractPhotos(taxonInfo: JsonObject?): List<JsonObject> {
        if (taxonInfo == null) return emptyList()
        val taxonPhotos = taxonInfo["taxon_photos"]
        if (taxonPhotos != null && taxonPhotos !is JsonNull && taxonPhotos is JsonArray) {
            return taxonPhotos.mapNotNull { element ->
                if (element is JsonNull) return@mapNotNull null
                val obj = try { element.jsonObject } catch (_: Exception) { return@mapNotNull null }
                val photo = obj["photo"]
                if (photo != null && photo !is JsonNull) {
                    try { photo.jsonObject } catch (_: Exception) { null }
                } else null
            }
        }
        val defaultPhoto = taxonInfo["default_photo"]
        if (defaultPhoto != null && defaultPhoto !is JsonNull) {
            return try { listOf(defaultPhoto.jsonObject) } catch (_: Exception) { emptyList() }
        }
        return emptyList()
    }

    private fun extractAncestorByRank(taxonInfo: JsonObject?, rank: String): String? {
        val ancestors = taxonInfo?.get("ancestors")
        if (ancestors == null || ancestors is JsonNull) return null
        val array = try { ancestors.jsonArray } catch (_: Exception) { return null }
        for (a in array) {
            if (a is JsonNull) continue
            val obj = try { a.jsonObject } catch (_: Exception) { continue }
            if (obj["rank"]?.jsonPrimitive?.contentOrNull == rank) {
                return obj["preferred_common_name"]?.jsonPrimitive?.contentOrNull
                    ?: obj["name"]?.jsonPrimitive?.contentOrNull
            }
        }
        return null
    }

    private fun extractAncestorScientificByRank(taxonInfo: JsonObject?, rank: String): String? {
        val ancestors = taxonInfo?.get("ancestors")
        if (ancestors == null || ancestors is JsonNull) return null
        val array = try { ancestors.jsonArray } catch (_: Exception) { return null }
        for (a in array) {
            if (a is JsonNull) continue
            val obj = try { a.jsonObject } catch (_: Exception) { continue }
            if (obj["rank"]?.jsonPrimitive?.contentOrNull == rank) {
                return obj["name"]?.jsonPrimitive?.contentOrNull
            }
        }
        return null
    }
}

File: ios/Manakin/Data/Generator/DatasetGenerator.swift
import Foundation

enum GenerationPhase: String {
    case fetchingSpecies = "FETCHING_SPECIES"
    case fetchingHistograms = "FETCHING_HISTOGRAMS"
    case fetchingDetails = "FETCHING_DETAILS"
    case downloadingPhotos = "DOWNLOADING_PHOTOS"
    case saving = "SAVING"
}

struct GenerationProgress {
    let phase: GenerationPhase
    let current: Int
    let total: Int
    let message: String
}

class DatasetGenerator {

    private let apiClient: INatApiClient

    private let encoder: JSONEncoder = {
        let e = JSONEncoder()
        e.outputFormatting = [.sortedKeys]
        return e
    }()

    private var datasetsDir: URL {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        return docs.appendingPathComponent("datasets")
    }

    init(apiClient: INatApiClient) {
        self.apiClient = apiClient
    }

    func generate(
        placeIds: [Int],
        placeName: String,
        taxonIds: [Int?],
        taxonName: String,
        groupName: String,
        minObs: Int = 1,
        qualityGrade: String = "research",
        maxPhotos: Int = 3
    ) -> AsyncThrowingStream<GenerationProgress, Error> {
        AsyncThrowingStream { continuation in
            Task {
                do {
                    let result = try await self.performGeneration(
                        placeIds: placeIds,
                        placeName: placeName,
                        taxonIds: taxonIds,
                        taxonName: taxonName,
                        groupName: groupName,
                        minObs: minObs,
                        qualityGrade: qualityGrade,
                        maxPhotos: maxPhotos,
                        onProgress: { progress in
                            continuation.yield(progress)
                        }
                    )
                    // Final progress indicating completion
                    continuation.yield(result)
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
        }
    }

    private func performGeneration(
        placeIds: [Int],
        placeName: String,
        taxonIds: [Int?],
        taxonName: String,
        groupName: String,
        minObs: Int,
        qualityGrade: String,
        maxPhotos: Int,
        onProgress: @Sendable (GenerationProgress) -> Void
    ) async throws -> GenerationProgress {
        let slug = "\(groupName)-\(placeName)"
            .lowercased()
            .replacingOccurrences(of: "[^a-z0-9]+", with: "-", options: .regularExpression)
            .trimmingCharacters(in: CharacterSet(charactersIn: "-"))

        let outputDir = datasetsDir.appendingPathComponent(slug)
        let photosDir = outputDir.appendingPathComponent("photos")
        try FileManager.default.createDirectory(at: photosDir, withIntermediateDirectories: true)

        // Step 1: Fetch species list across all taxon+place combos
        onProgress(GenerationProgress(phase: .fetchingSpecies, current: 0, total: 1, message: "Fetching species list..."))

        var speciesByTaxonId: [Int: [String: Any]] = [:]
        var obsCounts: [Int: Int] = [:]

        for taxonId in taxonIds {
            for placeId in placeIds {
                var page = 1
                while true {
                    let (_, results) = try await apiClient.getSpeciesCounts(
                        taxonId: taxonId,
                        placeId: placeId,
                        qualityGrade: qualityGrade,
                        page: page
                    )
                    for r in results {
                        guard let taxon = r["taxon"] as? [String: Any],
                              let tid = taxon["id"] as? Int else { continue }
                        let count = r["count"] as? Int ?? 0
                        if speciesByTaxonId[tid] == nil {
                            speciesByTaxonId[tid] = r
                            obsCounts[tid] = count
                        } else {
                            obsCounts[tid] = (obsCounts[tid] ?? 0) + count
                        }
                    }
                    onProgress(GenerationProgress(
                        phase: .fetchingSpecies,
                        current: speciesByTaxonId.count,
                        total: speciesByTaxonId.count,
                        message: "Found \(speciesByTaxonId.count) species so far..."
                    ))
                    if results.count < 200 || results.isEmpty { break }
                    page += 1
                }
            }
        }

        // Filter by min obs (using merged counts)
        let filteredIds = obsCounts.filter { $0.value >= minObs }.keys
        let filteredRaw: [[String: Any]] = filteredIds.compactMap { tid in
            guard var obj = speciesByTaxonId[tid] else { return nil }
            obj["count"] = obsCounts[tid] ?? 0
            return obj
        }

        let speciesList = DataProcessor.buildSpeciesList(rawCounts: filteredRaw)
        guard !speciesList.isEmpty else {
            throw GenerationError.noSpeciesFound
        }

        let nSpecies = speciesList.count
        onProgress(GenerationProgress(phase: .fetchingSpecies, current: nSpecies, total: nSpecies, message: "\(nSpecies) species found"))

        // Step 2: Fetch histograms - sum across all places
        var histograms: [String: [Int: Int]] = [:]
        for (i, sp) in speciesList.enumerated() {
            onProgress(GenerationProgress(
                phase: .fetchingHistograms,
                current: i + 1,
                total: nSpecies,
                message: "Histogram: \(sp.commonName.isEmpty ? sp.scientificName : sp.commonName) (\(i + 1)/\(nSpecies))"
            ))

            var combined: [Int: Int] = [:]
            for placeId in placeIds {
                let h = try await apiClient.getHistogram(taxonId: sp.taxonId, placeId: placeId, qualityGrade: qualityGrade)
                for (week, count) in h {
                    combined[week] = (combined[week] ?? 0) + count
                }
            }
            histograms[sp.scientificName] = combined
        }

        let weekly = DataProcessor.buildWeeklyMatrix(histograms: histograms)

        // Update peak/first/last week
        for sp in speciesList {
            guard let h = histograms[sp.scientificName] else { continue }
            let active = h.filter { $0.value > 0 }
            if !active.isEmpty {
                sp.peakWeek = active.max(by: { $0.value < $1.value })!.key
                sp.firstWeek = active.keys.min()!
                sp.lastWeek = active.keys.max()!
            }
            sp.periodCount = DataProcessor.detectFlightPeriods(weeklyData: weekly, speciesName: sp.scientificName)
        }

        // Step 3: Fetch taxa details
        onProgress(GenerationProgress(phase: .fetchingDetails, current: 0, total: nSpecies, message: "Fetching species details..."))
        var taxaDetails: [Int: [String: Any]] = [:]
        let allTaxonIds = speciesList.map { $0.taxonId }
        for i in stride(from: 0, to: allTaxonIds.count, by: 30) {
            let end = min(i + 30, allTaxonIds.count)
            let batch = Array(allTaxonIds[i..<end])
            let results = try await apiClient.getTaxaDetails(taxonIds: batch)
            for t in results {
                guard let id = t["id"] as? Int else { continue }
                taxaDetails[id] = t
            }
            onProgress(GenerationProgress(
                phase: .fetchingDetails,
                current: min(i + 30, allTaxonIds.count),
                total: nSpecies,
                message: "Details: \(min(i + 30, allTaxonIds.count))/\(nSpecies)"
            ))
        }

        // Step 4: Download photos
        var photoMap: [Int: [SpeciesPhoto]] = [:]
        for (idx, sp) in speciesList.enumerated() {
            onProgress(GenerationProgress(
                phase: .downloadingPhotos,
                current: idx + 1,
                total: nSpecies,
                message: "Photos: \(sp.commonName.isEmpty ? sp.scientificName : sp.commonName) (\(idx + 1)/\(nSpecies))"
            ))

            let taxonInfo = taxaDetails[sp.taxonId]
            let photos = extractPhotos(taxonInfo: taxonInfo)
            var spPhotos: [SpeciesPhoto] = []

            // Filter to Creative Commons licensed photos only
            let ccPhotos = photos.filter { photo in
                guard let license = photo["license_code"] as? String else { return false }
                return license.hasPrefix("cc")
            }

            // Fallback to observation photos if taxon photos are insufficient
            let allCcPhotos: [[String: Any]]
            if ccPhotos.count < maxPhotos {
                let obsPhotos = (try? await apiClient.getObservationPhotos(
                    taxonId: sp.taxonId,
                    placeIds: placeIds,
                    qualityGrade: qualityGrade,
                    maxResults: maxPhotos - ccPhotos.count
                )) ?? []
                allCcPhotos = ccPhotos + obsPhotos
            } else {
                allCcPhotos = ccPhotos
            }

            for (pi, photo) in allCcPhotos.prefix(maxPhotos).enumerated() {
                let url = photo["medium_url"] as? String ?? photo["url"] as? String
                guard var photoUrl = url else { continue }
                photoUrl = photoUrl.replacingOccurrences(of: "/square.", with: "/medium.")

                let filename = "\(sp.taxonId)_\(pi).jpg"
                guard let bytes = await apiClient.downloadPhoto(url: photoUrl) else { continue }

                let fileURL = photosDir.appendingPathComponent(filename)
                try? bytes.write(to: fileURL)

                spPhotos.append(SpeciesPhoto(
                    file: filename,
                    attribution: photo["attribution"] as? String,
                    license: photo["license_code"] as? String
                ))
            }
            photoMap[sp.taxonId] = spPhotos
        }

        // Step 5: Build and save dataset
        onProgress(GenerationProgress(phase: .saving, current: 0, total: 1, message: "Saving dataset..."))

        var weeklyBySpecies: [String: [WeeklyEntry]] = [:]
        for row in weekly {
            weeklyBySpecies[row.species, default: []].append(WeeklyEntry(week: row.week, n: row.n, relAbundance: row.relAbundance))
        }

        let speciesEntries: [Species] = speciesList.map { sp in
            let taxonInfo = taxaDetails[sp.taxonId]
            let family = extractAncestorByRank(taxonInfo: taxonInfo, rank: "family")
            let familySci = extractAncestorScientificByRank(taxonInfo: taxonInfo, rank: "family")
            let order = extractAncestorByRank(taxonInfo: taxonInfo, rank: "order")
            let orderSci = extractAncestorScientificByRank(taxonInfo: taxonInfo, rank: "order")

            let conservation = taxonInfo?["conservation_status"] as? [String: Any]
            var description = taxonInfo?["wikipedia_summary"] as? String ?? ""
            description = description.replacingOccurrences(of: "<[^>]+>", with: "", options: .regularExpression).trimmingCharacters(in: .whitespaces)

            var spWeekly = weeklyBySpecies[sp.scientificName] ?? []
            let existingWeeks = Set(spWeekly.map { $0.week })
            for w in 1...53 {
                if !existingWeeks.contains(w) {
                    spWeekly.append(WeeklyEntry(week: w, n: 0, relAbundance: 0))
                }
            }
            spWeekly.sort { $0.week < $1.week }

            return Species(
                taxonId: sp.taxonId,
                scientificName: sp.scientificName,
                commonName: sp.commonName,
                totalObs: sp.totalObs,
                rarity: sp.rarity,
                peakWeek: sp.peakWeek,
                firstWeek: sp.firstWeek,
                lastWeek: sp.lastWeek,
                periodCount: sp.periodCount,
                description: description,
                conservationStatus: conservation?["status"] as? String,
                conservationStatusName: conservation?["status_name"] as? String,
                family: family,
                familyScientific: familySci,
                order: order,
                orderScientific: orderSci,
                photos: photoMap[sp.taxonId] ?? [],
                weekly: spWeekly
            )
        }

        let dataset = Dataset(
            metadata: DatasetMetadata(
                placeName: placeName,
                placeId: placeIds.first ?? 0,
                placeIds: placeIds,
                group: groupName,
                taxonName: taxonName,
                taxonIds: taxonIds.compactMap { $0 },
                totalObs: speciesList.reduce(0) { $0 + $1.totalObs },
                speciesCount: speciesList.count,
                generatedAt: ISO8601DateFormatter().string(from: Date()),
                minObs: minObs,
                qualityGrade: qualityGrade,
                maxPhotos: maxPhotos
            ),
            species: speciesEntries
        )

        let jsonData = try encoder.encode(dataset)
        try jsonData.write(to: outputDir.appendingPathComponent("dataset.json"))

        let finalProgress = GenerationProgress(
            phase: .saving,
            current: 1,
            total: 1,
            message: "Complete! \(speciesEntries.count) species"
        )
        return finalProgress
    }

    // MARK: - Helpers

    private func extractPhotos(taxonInfo: [String: Any]?) -> [[String: Any]] {
        guard let taxonInfo = taxonInfo else { return [] }

        if let taxonPhotos = taxonInfo["taxon_photos"] as? [[String: Any]] {
            let photos = taxonPhotos.compactMap { element -> [String: Any]? in
                return element["photo"] as? [String: Any]
            }
            if !photos.isEmpty { return photos }
        }

        if let defaultPhoto = taxonInfo["default_photo"] as? [String: Any] {
            return [defaultPhoto]
        }

        return []
    }

    private func extractAncestorByRank(taxonInfo: [String: Any]?, rank: String) -> String? {
        guard let ancestors = taxonInfo?["ancestors"] as? [[String: Any]] else { return nil }
        for a in ancestors {
            if a["rank"] as? String == rank {
                return a["preferred_common_name"] as? String ?? a["name"] as? String
            }
        }
        return nil
    }

    private func extractAncestorScientificByRank(taxonInfo: [String: Any]?, rank: String) -> String? {
        guard let ancestors = taxonInfo?["ancestors"] as? [[String: Any]] else { return nil }
        for a in ancestors {
            if a["rank"] as? String == rank {
                return a["name"] as? String
            }
        }
        return nil
    }
}

enum GenerationError: Error, LocalizedError {
    case noSpeciesFound

    var errorDescription: String? {
        switch self {
        case .noSpeciesFound:
            return "No species found matching criteria."
        }
    }
}
